File: packages/python-sdk/e2b/sandbox/payments.py
import re
from dataclasses import dataclass
from typing import Dict, Literal, Optional


_PRIVATE_KEY_RE = re.compile(r"^0x[0-9a-fA-F]{64}$")
_KNOWN_NETWORKS = ("base", "base-sepolia")


@dataclass
class PaymentConfig:
    private_key: str
    network: Literal["base", "base-sepolia"] = "base-sepolia"
    max_spend: Optional[float] = None

    def __post_init__(self) -> None:
        if not _PRIVATE_KEY_RE.match(self.private_key):
            raise ValueError(
                "private_key must be a 0x-prefixed 32-byte hex string (64 hex chars after 0x)"
            )
        if self.network not in _KNOWN_NETWORKS:
            raise ValueError(
                f'network must be "base" or "base-sepolia", got "{self.network}"'
            )
        if self.max_spend is not None and self.max_spend < 0:
            raise ValueError("max_spend must be non-negative")


@dataclass
class PaymentEvent:
    timestamp: str
    url: str
    amount_usd: float
    tx_hash: str
    status: Literal["success", "failed"]


@dataclass
class SandboxPaymentsBalance:
    usdc: float
    address: str


def build_payment_envs(config: PaymentConfig) -> Dict[str, str]:
    envs: Dict[str, str] = {
        "E2B_PAYMENT_PRIVATE_KEY": config.private_key,
        "E2B_PAYMENT_NETWORK": config.network,
    }
    if config.max_spend is not None:
        envs["E2B_PAYMENT_MAX_SPEND"] = str(config.max_spend)
    return envs

File: packages/js-sdk/src/sandbox/payments.ts
export interface PaymentConfig {
  /** 0x-prefixed 32-byte hex private key for the wallet that pays. */
  privateKey: string
  /** Base network to use. Defaults to 'base-sepolia' (testnet). */
  network?: 'base' | 'base-sepolia'
  /** Optional USD spending cap. Payments exceeding this cumulatively are rejected inside the sandbox. */
  maxSpend?: number
}

export interface PaymentEvent {
  timestamp: string
  url: string
  amountUsd: number
  txHash: string
  status: 'success' | 'failed'
}

export interface SandboxPaymentsBalance {
  usdc: number
  address: string
}

const PRIVATE_KEY_RE = /^0x[0-9a-fA-F]{64}$/

export function validatePaymentConfig(config: PaymentConfig): void {
  if (!PRIVATE_KEY_RE.test(config.privateKey)) {
    throw new Error(
      'payments.privateKey must be a 0x-prefixed 32-byte hex string (64 hex chars after 0x)'
    )
  }
  if (
    config.network !== undefined &&
    config.network !== 'base' &&
    config.network !== 'base-sepolia'
  ) {
    throw new Error(
      `payments.network must be "base" or "base-sepolia", got "${config.network}"`
    )
  }
  if (config.maxSpend !== undefined && config.maxSpend < 0) {
    throw new Error('payments.maxSpend must be non-negative')
  }
}

export function buildPaymentEnvs(
  config: PaymentConfig
): Record<string, string> {
  const envs: Record<string, string> = {
    E2B_PAYMENT_PRIVATE_KEY: config.privateKey,
    E2B_PAYMENT_NETWORK: config.network ?? 'base-sepolia',
  }
  if (config.maxSpend !== undefined) {
    envs['E2B_PAYMENT_MAX_SPEND'] = config.maxSpend.toString()
  }
  return envs
}
